airlock/policies.py

```diff
@@ -0,0 +1,125 @@
+"""
+Check if a user is allowed to perform a particular action on a request or
+request file. This is dependent on our policies regarding changes that can
+be made:
+- while requests are in particular statuses
+- for types of file
+- at a certain phase in a request turn
+
+For example, a user has permission to review a particular request and vote on
+a file if they are an output checker and they are not the author of this request.
+However, in addition, the file must be of type OUTPUT, and the request must be
+in an editing status (submitted/partially reviewed/reviewed)
+"""
+
+from pathlib import Path
+from typing import TYPE_CHECKING
+
+from airlock import exceptions
+from airlock.permissions import check_user_can_action_request_for_workspace
+from airlock.types import UrlPath
+from airlock.users import User
+from airlock.utils import is_valid_file_type
+
+
+if TYPE_CHECKING:  # pragma: no cover
+    # We are avoiding circular dependencies by using forward references for
+    # type annotations where necessary, and this check so that type-checkin
+    # imports are not executed at runtime.
+    # https://peps.python.org/pep-0484/#forward-references
+    # https://mypy.readthedocs.io/en/stable/runtime_troubles.html#import-cycles`
+    from airlock.business_logic import ReleaseRequest, Workspace
+
+
+def check_user_can_edit_request(user: User, request: "ReleaseRequest"):
+    """
+    This user has permission to edit the request, AND the request is in an
+    editable state
+    """
+    if user.username != request.author:
+        raise exceptions.RequestPermissionDenied(
+            f"only author {request.author} can modify the files in this request"
+        )
+    if not request.is_editing():
+        raise exceptions.RequestPermissionDenied(
+            f"cannot modify files in request that is in state {request.status.name}"
+        )
```

Comment:
see discussion in slack - I guess I was thinking that half of this would go in policies & half in permissions, maybe that doesn't make life any easier though.